tsdb/head.go

```diff
@@ -1211,6 +1245,75 @@ func (a *headAppender) AddFast(ref uint64, t int64, v float64) error {
 	return nil
 }
 
+func (a *headAppender) AddExemplar(lset labels.Labels, t int64, e exemplar.Exemplar) error {
+	if t < a.minValidTime {
+		a.head.metrics.outOfBoundSamples.Inc()
+		return storage.ErrOutOfBounds
+	}
+
+	// Ensure no empty labels have gotten through.
+	lset = lset.WithoutEmpty()
+
+	if len(lset) == 0 {
+		return errors.Wrap(ErrInvalidExemplar, "empty labelset")
+	}
+
+	if l, dup := lset.HasDuplicateLabelNames(); dup {
+		return errors.Wrap(ErrInvalidExemplar, fmt.Sprintf(`label name "%s" is not unique`, l))
+	}
+
+	s, created, err := a.head.getOrCreate(lset.Hash(), lset)
+	if err != nil {
+		return err
+	}
+
+	// in theory the series should never be created as a result of an AddExemplar call, I think
```

Comment:
If that is the case, then we should not be calling getOrCreate and call getByHash instead. If it does not exist, then return an error. Else we might silently miss some bugs here.